### ISSUE TITLE
Custom control's BindingContext gets overwritten by MainPage's BindingContext

### ISSUE THREAD

**Issue by ToolmakerSteve** [taublast/DrawnUi.Maui#92]
How set separate BindingContext on a custom SkiaControl, when there is a different BindingContext on MainPage (in which that custom control is used)?

I haven't looked at Maui source code, but I assume Maui uses same BindingContext paradigm as Xamarin Forms: each control can set its own BindingContext.  
If left null, then a search is done up the view hierarchy.

DrawnUI.DrawnView interferes with this paradigm.  
its `OnBindingContextChanged()` recurses down through subviews, setting their BindingContext.

This wipes out any BindingContext set on those controls.

I could work around this, by making a subclass of SkiaLayout that overrides `ApplyBindingContext`. Put my custom control inside that.  
**Is there a cleaner fix available?**

-----------------------------
Thinking about how one could propagate BindingContext down, while respecting different values.  
An idea: add parameters:
```cs
public virtual void ApplyBindingContext(object oldValue, object newValue)
{
  foreach (var child in this.Views)
  {   // If child has a different BindingContext, leave it alone. But do replace null, for efficient access.
    if (child.BindingContext == null || child.BindingContext == oldValue)
      child.BindingContext = newValue;
  }
```

**Comment by ToolmakerSteve** [comment]
Found an acceptable hack to work around this:

In custom control:
```cs
// in constructor
  ...
  BindingContext = FrozenBindingContext = this;
  ...

// override
protected override void OnBindingContextChanged()
{
    if (FrozenBindingContext != null && BindingContext != FrozenBindingContext)
        BindingContext = FrozenBindingContext;

    base.OnBindingContextChanged();
}

// data
private object FrozenBindingContext;
```

The result is when `ApplyBindingContext` recursion changes control's BindingContext,  
it is forced back to correct value.

--------------------------------------
A better fix would be to change how BindingContext is set and propagated, in `SkiaControl.Shared.cs`.  
But that seemed to require changing all DrawnUI overrides of `ApplyBindingContext`.  
Possibly also overrides of `OnBindingContextChanged`.

**Comment by ToolmakerSteve** [comment]
Brainstorming. A way to distinguish between BindingContext set "directly" on a control, vs "inherited" context:

```cs
public partial class SkiaControl : VisualElement
{
    public object DirectBindingContext;

    public virtual void ApplyBindingContext()
    {
        // Whenever there is a DirectBindingContext, propagate that.
        object context = DirectBindingContext ?? BindingContext;
        
        foreach (var content in this.Views)
        {
             content.BindingContext = context;
        }

        foreach (var content in this.VisualEffects)
        {
            content.Attach(this);
        }

        if (FillGradient != null)
            FillGradient.BindingContext = context;
    }
}
```

 In subclasses that wish to set their own BindingContext:
```cs
public partial class SubclassNameHere
{
    ...
    public SubclassNameHere()
    {
        ...
        //was: BindingContext = ...;
        DirectBindingContext = ...;
        ...
    }
}
```

All code that sets "BindingContext" would have to be examined. To determine whether it should instead set `DirectBindingContext`

"client" code (an app using this framework) would set `DirectBindingContext`, to avoid their context being overwritten by an ancestor.

**Comment by taublast** [comment]
This is a very interesting subject. 

* The MAUI way of setting the binding context:

When the parent was set, adopt its binding context - call an **internal** virtual `SetChildInheritedBindingContext`, since this is internal like most of the MAUI mechanics we cannot use it. When the parent is lost (parent becomes null) sets binding context to null.

* DrawnUI way:

When the binding context changes then propagate binding context to children. This leads to a situation when a child cannot affect this, only the parent can, by overriding a ***public***  `ApplyBindingContext`. This method is similar to MAUI internal `SetInheritedBindingContext`.  At the same time when the parent is lost (parent becomes null) we set binding context to null.

This led, for example, to a hacky workaround for SkiaScroll, to avoid setting its biding context to a child when it already has its own:

```
   public override void ApplyBindingContext()
   {
       base.ApplyBindingContext();

       if (this.Content != null && Content.BindingContext == null) //todo remove this last condition!
       {
           Content.BindingContext = BindingContext;
       }
   }
```

So basically when this was all brought to light, thanks to You, I would tend to solve that by implementing similar to MAUI  **protected*** `SetChildInheritedBindingContext` alternative that would be invoked from `OnParentChanged`, while we wouldn't propagate context to children anymore. This way every control could implement it's own logic to accept or deny a new binding context.

Thoughts?

**Comment by ToolmakerSteve** [comment]
`SetInheritedBindingContext` would be called on the child?

Yes, that would work. I would consider also adding a `public virtual SetBindingContext`, for situations where the goal is to tell an instance it should definitely use this new context. But still give it a chance to react to the change. 

Train ourselves and anyone who uses DrawnUI to avoid doing "BindingContext = ..." directly. Instead call one of these two methods. Then all classes could override, to change how they handle the two different requests.

Default behavior something like this:
```cs
// Use ourBindingContext. If it is null, use inheritedBindingContext.
protected object ourBindingContext;
protected object inheritedBindingContext;
protected virtual void SetBindingContext(object context)
{
	if (ourBindingContext != context)
	{
		ourBindingContext = context;
		// If ours gets cleared, use inherited one.
		if (context == null)
			BindingContext = inheritedBindingContext;
	}
}
protected virtual void SetInheritedBindingContext(object context)
{
	if (inheritedBindingContext != context)
	{
		inheritedBindingContext = context;
		// Only use inherited context if ours is null.
		if (ourBindingContext == null)
			BindingContext = context;
	}
}

```

**Comment by taublast** [comment]
So the fix would be illustrated with new code:

For SkiaScroll compared to the old code above, the new code is:

```
        public override void ApplyBindingContext()
        {
            base.ApplyBindingContext();

            Content?.SetInheritedBindingContext(BindingContext);
        }
```
And in the base SkiaControl:

```
        /// <summary>
        /// This is to be called by layouts to propagate their binding context to children. By overriding this method any child could deny a new context or use any other custom logic. To force new context for child parent would set child's BindingContext directly skipping the use of this method.
        /// </summary>
        /// <param name="context"></param>
        public virtual void SetInheritedBindingContext(object context)
        {
            BindingContext ??= context; //only if existing is null
        }
```

Will push soon

**Comment by taublast** [comment]
Closing with 34425e65ec5fd121300942722364f5ccbc10d8d0 will be in the next nuget

**Comment by taublast** [comment]
Just PR'd to main branch, you can already use it, while I guess we'd need a small bit more a new nuget

**Comment by taublast** [comment]
Will be affected by 6cb18faa35c857b9913c3d41ba05182e2a502507

**Comment by taublast** [comment]
> When a parent has children attached it sets their binding content to its own by calling `SetInheritedBindingContext` of the child ONLY if child's BindingContext is actually null. So when the Parent property of the child gets set to null this child BindingContext is set to null too.   
> 
> Any `SkiaControl` implements `public virtual void SetInheritedBindingContext(object context)` that you can override to prohibit changing binding context or apply any other logic.
> 
> Another case is when the control `BindingContext` changes for any reason, `public virtual void ApplyBindingContext()` method is invoked. Layouts override it to implement additional logic to modify or not its children `BindingContext`. Some view containers use the following logic:
